Export: Add timestamp injection, path and namespace support.

EreMaijala opened the pull request:
@demiankatz Could you check if these work for you? Docs missing, but you can do e.g.

`./console re:ex --source=foo --inject-created='created' --inject-date="updated" foo.xml`

or 

`./console re:ex --source=foo --inject-created='custom:date/created[@foo="bar"]' --inject-date="custom:date/updated" --add-namespace="custom=http://foobar" foo.xml`

EreMaijala committed:
Export: Add timestamp injection, path and namespace support.

demiankatz commented on src/RecordManager/Base/Command/Records/Export.php:
```diff
@@ -716,4 +795,134 @@ protected function writeRecord(string $record): void
+            // Extract node name and attributes:
+            $parts = explode('[', $pathPart);
+            $nodeParts = explode(':', $parts[0], 2);
+            if (isset($nodeParts[1])) {
```
If you want to reduce number of lines of code, maybe:

```
if (count($nodeParts) < 2) {
    array_unshift($nodeParts, null);
}
[$prefix, $nodeName] = $nodeParts;
```

(Not that there's anything wrong with the other way...)

EreMaijala replied:
I decided to ignore this to avoid `count` and array manipulation for imagined performance advantage. :)

demiankatz commented on src/RecordManager/Base/Command/Records/Export.php:
```diff
@@ -716,4 +795,134 @@ protected function writeRecord(string $record): void
+     * Add an node to the XML record
+     *
+     * @param \SimpleXMLElement $xml       XML record
+     * @param array             $path      XML path
```
Maybe add "from parseXmlPath()" to the doc comment to clarify what kind of array we're working with here.

demiankatz commented on src/RecordManager/Base/Command/Records/Export.php:
```diff
@@ -716,4 +795,134 @@ protected function writeRecord(string $record): void
+     */
+    protected function getXmlNamespaceFromPrefix(\SimpleXMLElement $xml, ?string $prefix): ?string
+    {
+        $ns = null;
```
Do you really need the `$ns` variable? Can you just return results directly instead of assigning to $ns, and return null as a default at the end?

EreMaijala replied:
Yes, and I simplified it further since it makes more sense to use as default whatever is parent node's default namespace instead of document's first namespace.

demiankatz reviewed (commented):
Thanks, @EreMaijala, I think this looks fine. I haven't done hands-on testing of every permutation, but the tests I tried were successful. Obviously, some of this might benefit from unit test coverage at some point, but I understand if time does not permit.

I don't think any changes are necessary, but I suggested a few possible optimizations/improvements below, which you can use or ignore as you see fit. :-)

EreMaijala committed:
Add namespace prefixes to DOM, tweak code and parameter descriptions.

EreMaijala commented:
@demiankatz Tweaked and simplified. Does it still work for you? I changed the namespace handling a bit so that if you add an element like `customNS:foo/bar`, `bar` will inherit `customNS`.

demiankatz commented:
@EreMaijala, I just exported again and got the same results, so I think all is well. I'll review the code changes later when I have a bit more time!

demiankatz commented on src/RecordManager/Base/Command/Records/Export.php:
```diff
@@ -492,23 +501,28 @@ protected function configure()
                 null,
                 InputOption::VALUE_REQUIRED,
-                'Inject last internal update timestamp of the record to the given XML field (ISO 8601 format)'
+                'Inject last internal update timestamp of the record to'
```
I think there's a stray "to" here.

EreMaijala replied:
Fixed!

demiankatz reviewed (changes requested):
I think we're nearly there -- just one typo to address. :-)

EreMaijala committed:
Fix typo.

demiankatz reviewed (approved):
Thanks, @EreMaijala!
